### ISSUE TITLE
Crash in writeCopyTo

### ISSUE THREAD

**Issue by sipersso** [realm/realm-swift#7702]
### How frequently does the bug occur?

Sometimes

### Description

I have seen crashes in production when calling realm.writeCopyTo (see stacktrace).

This started happening in Realm 10.22.0 (which had a lot of issues), but I have since then upgraded to 10.24.0 and I am still seeing some crashes. 

In this case it looks like there might be a loop in do_write_deep? The crash is repeated for the users affected.

### Stacktrace & log output

```shell
Crashed: com.apple.main-thread
0  MyApp                       0x42db08 realm::Array::init_from_ref(unsigned long) + 573 (alloc.hpp:573)
1  MyApp                       0x42e120 realm::Array::do_write_deep(realm::_impl::ArrayWriterBase&, bool) const + 843 (array.hpp:843)
2  MyApp                       0x42e138 realm::Array::do_write_deep(realm::_impl::ArrayWriterBase&, bool) const + 846 (array.hpp:846)
3  MyApp                       0x42e138 realm::Array::do_write_deep(realm::_impl::ArrayWriterBase&, bool) const + 846 (array.hpp:846)
4  MyApp                       0x42e138 realm::Array::do_write_deep(realm::_impl::ArrayWriterBase&, bool) const + 846 (array.hpp:846)
5  MyApp                       0x42e138 realm::Array::do_write_deep(realm::_impl::ArrayWriterBase&, bool) const + 846 (array.hpp:846)
6  MyApp                       0x42e138 realm::Array::do_write_deep(realm::_impl::ArrayWriterBase&, bool) const + 846 (array.hpp:846)
7  MyApp                       0x42e138 realm::Array::do_write_deep(realm::_impl::ArrayWriterBase&, bool) const + 846 (array.hpp:846)
8  MyApp                       0x42e138 realm::Array::do_write_deep(realm::_impl::ArrayWriterBase&, bool) const + 846 (array.hpp:846)
9  MyApp                       0x42e138 realm::Array::do_write_deep(realm::_impl::ArrayWriterBase&, bool) const + 846 (array.hpp:846)
10 MyApp                       0x42e138 realm::Array::do_write_deep(realm::_impl::ArrayWriterBase&, bool) const + 846 (array.hpp:846)
11 MyApp                       0x477d60 realm::Group::write(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, int, realm::Group::TableWriter&, bool, bool, unsigned long long) + 1275 (group.cpp:1275)
12 MyApp                       0x478720 realm::Group::write(realm::util::File&, char const*, unsigned long long, realm::Group::TableWriter&) const + 167 (streambuf:167)
13 MyApp                       0x478948 realm::Group::write(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const*, unsigned long long, bool) const + 1016 (file.hpp:1016)
14 MyApp                       0x3fb270 -[RLMRealm writeCopyToURL:encryptionKey:error:] + 1462 (string:1462)
15 MyApp                       0x75b918 Realm.writeCopy(toFile:encryptionKey:) + 914 (Realm.swift:914)
```

### Can you reproduce the bug?

Not yet

### Reproduction Steps

This only happened in a few cases. I am not able to reproduce it easily

### Version

10.22.0, 10.24.0

### What SDK flavour are you using?

Local Database only

### Are you using encryption?

No, not using encryption

### Platform OS and version(s)

iOS 15

### Build environment

Xcode version: 13.2.1
Dependency manager and version: SPM

**Comment by sipersso** [comment]
let tmpFile = URL.init(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent("\(UUID().uuidString)")        
        autoreleasepool{
            let realm = try! Realm.init(configuration: realmConfiguration)
            try! realm.writeCopy(toFile: tmpFile)
        }
        
Not sure exactly what you need. The configuration passed in is just the Realm.defaultConfiguration. Not sure if the autoreleasepool could have anything to do with it? It only happens rarely, so I am not able to reproduce it.

**Comment by dianaafanador3** [comment]
@jedelbo Can you help with this?

**Comment by jedelbo** [comment]
I am not sure what the exact error is here. The line `alloc.hpp:573` does not contain any assertion or call to terminate. So I would wish that we had more information about the crash. Is this a segmentation fault, and assertion or something else? This error should not have anything to do with the write_copy feature as such, as it happens in `Group::write` which is some very old functionality.

**Comment by sipersso** [comment]
I just checked in firebase and noticed that I missed to include the actual line of the crash. Here is the actual error that happens:

Crashed: com.apple.main-thread
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x00000e0283a7ca60

@jedelbo does that help?

I have only seen this happen for one user so far, but it is repeated crashes.

**Comment by sipersso** [comment]
I now have gotten some more info since of the users actually getting this error has reached out. Apparently the call to writeCopyTo is done on the main thread. I assume that this can be an issue and that it should be moved to a background thread?

**Comment by pavel-ship-it** [comment]
@sipersso,
Yes, blocking the main thread definitely can be a reason. And this is certainly worth to try. If you have any more information please update the ticket.

**Comment by sipersso** [comment]
I actually have more information. WriteCopyTo fails, but I was able to get the file anyway by avoiding the Realm apis and extracting the realm file via the configuration path. It turns out that the file had corrupt data in it! 

I started getting all kinds of strange errors after the 10.22.0 update and now I am thinking that all of these are caused by corrupt data. I know there have been open issues here https://github.com/realm/realm-core/issues/5174.

For this user in particular, I was able to open the extracted realm file in realm studio and I could manually delete objects that had collections pointing to bad data, but I have seen the error happen in crash logs and unless users reach out and get a special build (that doesn't use any Realm API:s), I can't help them. 

If it helps, I have a support case in the MongoDb Atlas portal with case number 00900582 where I have uploaded the realm file I extracted. 

Not sure how to handle this. Opening a realm with corrupt data will not throw errors, as you use assertions to verify the correct type of database fields, so I can show the user an error message, it will just repeatedly crash without a way for them to reach out. 

I can't share all details here, but I can share more via the support portal or if I have an email I can reach you at.
